### PR TITLE
Add flag to decompress brotli server-side

### PR DESCRIPTION
**Pull request by raphjaph**

### CODE REVIEW

File: docs/src/guides/testing.md [hunk 1]
```diff
@@ -46,6 +46,10 @@ View the inscription in the regtest explorer:
 ```
 ord -r server
 ```
+If you want to test Brotli encoded content over http you need to add:
```
Comment (casey) [inline docs/src/guides/testing.md]:
```suggestion

By default, browsers don't support compression over HTTPS. To test compressed content over HTTP, add the `--decompress` flag:
```

Comment (raphjaph) [reply docs/src/guides/testing.md]: we're only decompressing brotli so made it explicit

**Review by casey — CHANGES_REQUESTED** [review]
Random comments.

File: src/subcommand/server.rs [hunk 2]
```diff
@@ -49,9 +50,10 @@ mod accept_encoding;
 mod accept_json;
 mod error;
 
-#[derive(Clone)]
+#[derive(Clone, Default)]
 pub struct ServerConfig {
```
[the remark this answers is not in the document]
Comment (raphjaph) [reply src/subcommand/server.rs:54]: I'll have a look if it makes sense

**Comment by raphjaph** [comment]
I combined `PageConfig` and `ServerConfig` inside `ServerConfig`

File: docs/src/guides/testing.md [hunk 3]
```diff
@@ -46,6 +46,11 @@ View the inscription in the regtest explorer:
 ```
 ord -r server
 ```
+By default, browsers don't support compression over HTTPS. To test compressed
```
Comment (casey) [inline docs/src/guides/testing.md:49]:
```suggestion

By default, browsers don't support compression over HTTPS. To test compressed
```

File: docs/src/guides/testing.md [hunk 4]
```diff
@@ -46,6 +46,11 @@ View the inscription in the regtest explorer:
...
+content over HTTP, add the `--decompress` flag:
```
Comment (casey) [inline docs/src/guides/testing.md:50]:
```suggestion
content over HTTP, use the `--decompress` flag:
```

File: src/server_config.rs [hunk 5]
```diff
@@ -1,9 +1,11 @@
 use super::*;
 
 #[derive(Clone, Default)]
```
Comment (casey) [inline src/server_config.rs:3]: I checked and we can remove this `Clone`:

```suggestion
#[derive(Default)]
```

File: src/server_config.rs [hunk 6]
```diff
@@ -1,9 +1,11 @@
...
-pub(crate) struct PageConfig {
+pub(crate) struct ServerConfig {
   pub(crate) chain: Chain,
   pub(crate) csp_origin: Option<String>,
+  pub(crate) decompress_brotli: bool,
```
Comment (casey) [inline src/server_config.rs:7]:
```suggestion
  pub(crate) decompress: bool,
```

File: src/subcommand/server.rs [hunk 7]
```diff
@@ -466,7 +465,7 @@ impl Server {
   }
 
   async fn sat(
-    Extension(page_config): Extension<Arc<PageConfig>>,
+    Extension(page_config): Extension<Arc<ServerConfig>>,
```
Comment (casey) [inline src/subcommand/server.rs:468]: We should change `page_config` to server config:

```suggestion
    Extension(server_config): Extension<Arc<ServerConfig>>,
```

There's also a bunch of other places we call it `page_config`, the easiest way to find them is probably just grepping.

**Comment by casey** [comment]
> Acceptability shouldn't be checked since browsers won't set it if connecting via HTTP

I don't think I understand. We can check it anyways, since if it isn't set, we can just decmpress.

File: src/subcommand/server.rs [hunk 8]
```diff
@@ -998,7 +997,8 @@ impl Server {
   async fn content(
     Extension(index): Extension<Arc<Index>>,
     Extension(config): Extension<Arc<Config>>,
-    Extension(page_config): Extension<Arc<PageConfig>>,
+    Extension(page_config): Extension<Arc<ServerConfig>>,
+    Extension(server_config): Extension<Arc<ServerConfig>>,
```
Comment (casey) [inline src/subcommand/server.rs:1001]:
```suggestion
    Extension(server_config): Extension<Arc<ServerConfig>>,
```

File: src/subcommand/server.rs [hunk 9]
```diff
@@ -1020,29 +1020,11 @@ impl Server {
   fn content_response(
     inscription: Inscription,
     accept_encoding: AcceptEncoding,
-    page_config: &PageConfig,
+    page_config: &ServerConfig,
+    server_config: &ServerConfig,
```
Comment (casey) [inline src/subcommand/server.rs:1024]:
```suggestion
    server_config: &ServerConfig,
```

File: src/subcommand/server.rs [hunk 10]
```diff
@@ -1116,7 +1136,7 @@ impl Server {
           .into_response(),
       ),
       Media::Iframe => Ok(
-        Self::content_response(inscription, accept_encoding, &page_config)?
+        Self::content_response(inscription, accept_encoding, &page_config, &server_config)?
```
Comment (casey) [inline src/subcommand/server.rs:1139]:
```suggestion
        Self::content_response(inscription, accept_encoding, &server_config)?
```

**Review by casey — CHANGES_REQUESTED** [review]

**Comment by raphjaph** [comment]
> > Acceptability shouldn't be checked since browsers won't set it if connecting via HTTP
> 
> I don't think I understand. We can check it anyways, since if it isn't set, we can just decmpress.

If we are decompressing the data will not be encoded so we shouldn't set the `CONTENT_ENCODING` header. Even if the client sets `br` in the request header we don't care since we are decompressing. But maybe I'm understanding something wrong.

File: src/subcommand/server.rs [hunk 11]
```diff
@@ -1068,6 +1048,43 @@ impl Server {
...
+
+    if let Some(content_encoding) = inscription.content_encoding() {
+      if accept_encoding.is_acceptable(&content_encoding) {
+        headers.insert(header::CONTENT_ENCODING, content_encoding);
+      } else if server_config.decompress
+        && content_encoding
+          .to_str()
+          .map_err(|err| ServerError::Internal(err.into()))?
+          .to_lowercase()
+          == "br"
```
Comment (casey) [inline src/subcommand/server.rs:1067]: I think we can just do:

```suggestion
        && content_encoding == "br"
```

**Review by casey — APPROVED** [review]
Very minor comment. Otherwise looks good.

**Comment by casey** [comment]
lgtm

**Pull request merged**